### PR TITLE
Add admin merge commands for nations and towns

### PR DESCRIPTION
**Pull request by Gamer153**
#### Description: 

I've added a command to ta so that admins can (force)merge towns and nations now.
____
#### New Nodes/Commands/ConfigOptions: 

`/ta <town|nation> <name> <merge|forcemerge> <other name>`

____
#### Relevant Towny Issue ticket:

#5489

____
- [x] I have tested this pull request for defects on a server. 

By making this pull request, I represent that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the TownyAdvanced organization has the copyright to use and modify my contribution under the Towny [License](https://github.com/LlmDl/Towny/blob/master/LICENSE.md) for perpetuity.

### CODE REVIEW

File: src/com/palmergames/bukkit/towny/command/TownCommand.java [hunk 1]
```diff
@@ -4056,16 +4061,16 @@ public static void parseTownMergeCommand(Player player, String[] args) {
...
 				TownyMessaging.sendMsg(player, Translatable.of("msg_town_merge_cancelled"));
 				return;
 			}).sendTo(player);
 		} else
-			sendTownMergeRequest(player, remainingTown, succumbingTown, cost);
+			sendTownMergeRequest(player, remainingTown, succumbingTown, cost, admin);
 	}
 
-	private static void sendTownMergeRequest(Player sender, Town remainingTown, Town succumbingTown, double cost) {
+	private static void sendTownMergeRequest(Player sender, Town remainingTown, Town succumbingTown, double cost, boolean admin) {
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/command/TownCommand.java:4073]: admin is unused here.

File: src/com/palmergames/bukkit/towny/command/TownCommand.java [hunk 2]
```diff
@@ -4056,16 +4061,16 @@ public static void parseTownMergeCommand(Player player, String[] args) {
...
 			Confirmation.runOnAccept(() -> {
-				sendTownMergeRequest(player, finalRemainingTown, finalSuccumbingTown, finalCost);
+				sendTownMergeRequest(player, finalRemainingTown, finalSuccumbingTown, finalCost, false);
 			}).runOnCancel(() -> {
 				TownyMessaging.sendMsg(player, Translatable.of("msg_town_merge_cancelled"));
 				return;
 			}).sendTo(player);
 		} else
-			sendTownMergeRequest(player, remainingTown, succumbingTown, cost);
+			sendTownMergeRequest(player, remainingTown, succumbingTown, cost, admin);
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/command/TownCommand.java:4070]: admin is never used.

File: src/com/palmergames/bukkit/towny/command/TownCommand.java [hunk 3]
```diff
@@ -4056,16 +4061,16 @@ public static void parseTownMergeCommand(Player player, String[] args) {
 			final Town finalRemainingTown = remainingTown;
 			final double finalCost = cost;
 			Confirmation.runOnAccept(() -> {
-				sendTownMergeRequest(player, finalRemainingTown, finalSuccumbingTown, finalCost);
+				sendTownMergeRequest(player, finalRemainingTown, finalSuccumbingTown, finalCost, false);
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/command/TownCommand.java:4064]: admin variable is never used.

File: src/com/palmergames/bukkit/towny/command/TownCommand.java [hunk 4]
```diff
@@ -4047,7 +4052,7 @@ public static void parseTownMergeCommand(Player player, String[] args) {
 			}			
 		}
 		
-		if (cost > 0) {
+		if (cost > 0 || admin) {
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/command/TownCommand.java:4055]: I dont think this `|| admin` is needed here, I think what you were going for was that admin-initiated merges wouldn't cost anything? What you probably want on line 4043 is `cost = admin ? 0 : baseCost + townblockCost + bankruptcyCost;` then the diff on 4045 can be undone.

File: src/com/palmergames/bukkit/towny/command/TownCommand.java [hunk 5]
```diff
@@ -3978,14 +3978,19 @@ public static void parseTownUnclaimCommand(Player player, String[] split) {
...
+
+	public static void parseTownMergeCommand(Player player, String[] args, boolean admin, @Nullable Town town) {
 		Resident resident = TownyUniverse.getInstance().getResident(player.getUniqueId());
-		if (!resident.isMayor()) {
+		if (!resident.isMayor() && !admin) {
 			TownyMessaging.sendErrorMsg(player, Translatable.of("msg_town_merge_err_mayor_only"));
 			return;
 		}
-		Town remainingTown = TownyAPI.getInstance().getResidentTownOrNull(resident); // The isMayor() test guarantees we get a town.		
+		if (admin && town == null) return;
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/command/TownCommand.java:3992]: This should return some feedback to the sender, unless we're certain that town will never be null when sent via admin.

**Review by LlmDl — CHANGES_REQUESTED** [review]
There's some changes to be made before I try testing this out.

**Review by Warriorrrr — COMMENTED** [review]
I don't think the forcemerge commands are really needed, the merge command should just do that. Having a confirmation would also be ideal.

**Comment by LlmDl** [comment]
> I don't think the forcemerge commands are really needed, the merge command should just do that. Having a confirmation would also be ideal.

people have brought up the merge commands dont work if the towns combined together will fail teh too-many-bought-townblocks tests.

File: src/com/palmergames/bukkit/towny/command/TownCommand.java (hunk 5 above)
Comment (Gamer153) [reply src/com/palmergames/bukkit/towny/command/TownCommand.java:3992]: I'm just gonna remove the admin variable, I can just use `town != null` instead.

**Comment by Gamer153** [comment]
The admin merge is supposed to be the same as if the mayor of the town sent a merge request to the other town, and the forcemerge is just supposed to merge the towns ignoring any checks.

File: src/com/palmergames/bukkit/towny/command/TownyAdminCommand.java [hunk 6]
```diff
@@ -1491,13 +1506,21 @@ public void parseAdminNationCommand(String[] split) throws TownyException {
 					TownyMessaging.sendErrorMsg(getSender(), Translatable.of("msg_invalid_name"));
 
 			} else if (split[1].equalsIgnoreCase("merge")) {
+				NationCommand.mergeNation(player, StringMgmt.remArgs(split, 1), nation);
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/command/TownyAdminCommand.java:1509]:
```suggestion
				NationCommand.mergeNation(player, StringMgmt.remArgs(split, 2), nation);

```

File: src/com/palmergames/bukkit/towny/command/TownyAdminCommand.java [hunk 7]
```diff
@@ -1342,6 +1349,14 @@ public void parseAdminTownCommand(String[] split) throws TownyException {
...
 			} else if (split[1].equalsIgnoreCase("trust")) {
 				TownCommand.parseTownTrustCommand(player, StringMgmt.remArgs(split, 2), town);
+			} else if (split[1].equalsIgnoreCase("merge")) {
+				TownCommand.parseTownMergeCommand(player, StringMgmt.remArgs(split, 2), town);
+			} else if (split[1].equalsIgnoreCase("forcemerge")) {
+				Town remainingTown = townyUniverse.getTown(split[2]);
+
+				if (remainingTown == null || remainingTown.equals(town))
+					throw new TownyException(Translatable.of("msg_err_invalid_name", split[2]));
+				Confirmation.runOnAccept(() -> townyUniverse.getDataSource().mergeTown(town, remainingTown)).sendTo(sender);
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/command/TownyAdminCommand.java:1359]:
```suggestion
				Confirmation.runOnAccept(() -> townyUniverse.getDataSource().mergeTown(town, remainingTown)).sendTo(sender);
				//TODO: Add language string after merge.

```

**Review by LlmDl — CHANGES_REQUESTED** [review]

File: src/com/palmergames/bukkit/towny/command/TownyAdminCommand.java (hunk 6 above)
Comment (Gamer153) [reply src/com/palmergames/bukkit/towny/command/TownyAdminCommand.java:1509]: The mergeNation method expects the arguments `["merge", "<name>"]`, thats why its needed to include the `"merge"` too.

**Review by LlmDl — APPROVED** [review]

**Review by LlmDl — APPROVED** [review]

**Pull request merged**